### ISSUE TITLE
Conditional Normalizing Flow: Allow parameters to vary as function of features

### ISSUE THREAD

**Issue by KiwiAthlete** [pyro-ppl/pyro#3240]
### Description
Thanks for the great package!

I was wondering how to use the Spline-Flow where the parameters of the model are allowed to vary with features. The idea would be to condition the learnable parameters of the Spline-Flow on a set of features. 

### Example Code
The example give in [Learnable Univariate Distributions in Pyro](https://pyro.ai/examples/normalizing_flows_i.html#Learnable-Univariate-Distributions-in-Pyro) and replicated for convenience below is an unconditional example, where the parameters are independent of any features.

```python
# Imports
import torch
import pyro
import pyro.distributions as dist
import pyro.distributions.transforms as T
from torch.utils.data import DataLoader
from sklearn import datasets
from sklearn.preprocessing import StandardScaler

# Create Data
torch.manual_seed(123)
n_samples = 1000
X, y = datasets.make_circles(n_samples=n_samples, factor=0.5, noise=0.05)
X = StandardScaler().fit_transform(X)
X = X[:,0].reshape(-1,1)

# Create Normalizing Flow
n_targets = 1
base_dist = dist.Normal(torch.zeros(n_targets), torch.ones(n_targets))
spline_transform = T.Spline(n_targets, count_bins=5)
flow_dist = dist.TransformedDistribution(base_dist, [spline_transform])

# Train Model
steps = 1000
dataset = torch.tensor(X, dtype=torch.float)
optimizer = torch.optim.Adam(spline_transform.parameters(), lr=1e-2)
for step in range(steps):
    optimizer.zero_grad()
    loss = -flow_dist.log_prob(dataset).mean()
    loss.backward()
    optimizer.step()
    flow_dist.clear_cache()

    if step % 200 == 0:
        print('step: {}, loss: {}'.format(step, loss.item()))

# Sample 
X_flow = flow_dist.sample(torch.Size([100,])).detach().numpy()
# X_flow.shape = [100,1]
```

Looking at the parameters of the Spline-Flow shows that they are all unconditional.

```python
parameters = spline_transform.parameters()
for param in parameters:
    print(param)

# Parameter containing:
# tensor([[ 0.7785, -0.7411,  0.9119, -0.1716, -1.2727]], requires_grad=True)
# Parameter containing:
# tensor([[-1.4670, -1.5930, -2.0247, -0.1197, -0.4249]], requires_grad=True)
# Parameter containing:
# tensor([[-0.6859, -0.8958,  0.8580,  0.2149]], requires_grad=True)
```
All of them are of shape `(1,-1)`, where `1` is due to the parameters being unconditional. However, given a feature `x_feat` of shape `(1000,1)` what I want is the shape of the parameters to be `(1000,-1)` so that for each observation, there is a different set of parameters. Also, drawing 100 samples from the conditional distribution should return a shape of `(1000, 100)`, since for each of the `1000` observations, there should be `100` samples. The unconditional example from above returns `flow_dist.sample(torch.Size([1000,])).detach().numpy().shape = (100,1)`. 

I know there is a [Conditional Transforms in Pyro example](https://pyro.ai/examples/normalizing_flows_i.html#Conditional-Transforms-in-Pyro), but that is not what I want.

Looking forward to your suggestions.

Many thanks!

**Comment by martinjankowiak** [comment]
what you're describing wouldn't be fundamentally different from a conditional normalizing flow---it'd just be a different choice of neural architecture that consumes features `x` in a somewhat different way. you're free to implement such a neural network if you want but unfortunately there are not sufficient developer resources for pyro to maintain every imaginal neural architecture for normalizing flows.

**Comment by KiwiAthlete** [comment]
Ok thanks, closing it.
